app/feature/feature-movingflow/src/main/kotlin/com/hedvig/android/feature/movingflow/ui/chosecoveragelevelanddeductible/ChoseCoverageLevelAndDeductibleViewModel.kt

```diff
@@ -0,0 +1,194 @@
+          }
+          .filterNotNull()
+          .map { moveHomeQuote ->
+            CoverageInfo(moveHomeQuote.id, moveHomeQuote.tierName, moveHomeQuote.tierDescription, moveHomeQuote.premium)
+          }
+        val selectedCoverage = initiallySelectedHomeQuote ?: homeQuotes.first()
+        tiersInfo = TiersInfo(
+          allOptions = homeQuotes,
+          coverageOptions = uniqueCoverageOptions,
+          selectedCoverage = selectedCoverage,
+          selectedDeductible = null,
+        ).some()
+      }
+    }
+
+    return when (val tiersInfoValue = tiersInfo) {
+      None -> ChoseCoverageLevelAndDeductibleUiState.Loading
+      is Some -> {
+        when (val state = tiersInfoValue.value) {
+          null -> ChoseCoverageLevelAndDeductibleUiState.MissingOngoingMovingFlow
+          else -> ChoseCoverageLevelAndDeductibleUiState.Content(state)
+        }
+      }
+    }
+  }
+}
+
+sealed interface ChoseCoverageLevelAndDeductibleEvent {
+  data class SelectCoverage(val homeQuoteId: String) : ChoseCoverageLevelAndDeductibleEvent
+
+  data class SelectDeductible(val homeQuoteId: String) : ChoseCoverageLevelAndDeductibleEvent
+}
+
+internal sealed interface ChoseCoverageLevelAndDeductibleUiState {
+  data object MissingOngoingMovingFlow : ChoseCoverageLevelAndDeductibleUiState
+
+  data object Loading : ChoseCoverageLevelAndDeductibleUiState
+
+  data class Content(
+    val tiersInfo: TiersInfo,
+  ) : ChoseCoverageLevelAndDeductibleUiState {
+    val canSubmit = tiersInfo.selectedHomeQuoteId != null
+  }
+}
+
+data class CoverageInfo(
+  val moveHomeQuoteId: String,
+  val tierName: String,
+  val tierDescription: String?,
+  val premium: UiMoney,
+)
+
+internal data class TiersInfo(
+  val allOptions: List<MoveHomeQuote>,
+  val coverageOptions: List<CoverageInfo>,
+  val selectedCoverage: MoveHomeQuote,
+  val selectedDeductible: MoveHomeQuote?,
+) {
+  val deductibleOptions: DeductibleOptions = allOptions.filter { homeQuote ->
+    homeQuote.tierLevel == selectedCoverage.tierLevel
+  }.let { moveHomeQuotes ->
+    if (moveHomeQuotes.size <= 1) {
+      val deductible = selectedCoverage.deductible
+      if (deductible == null) {
+        DeductibleOptions.NoOptions
+      } else {
+        DeductibleOptions.OneOption(DeductibleOption(selectedCoverage.id, selectedCoverage.premium, deductible))
+      }
+    } else {
+      val allOptionsWithDeductible = moveHomeQuotes.filter { it.deductible != null }
+      when (allOptionsWithDeductible.size) {
+        0 -> DeductibleOptions.NoOptions
+        1 -> {
+          val onlyOption = allOptionsWithDeductible.first()
+          DeductibleOptions.OneOption(DeductibleOption(onlyOption.id, onlyOption.premium, onlyOption.deductible!!))
+        }
+
+        else -> DeductibleOptions.MutlipleOptions(
+          allOptionsWithDeductible.map { moveHomeQuote ->
+            DeductibleOption(moveHomeQuote.id, moveHomeQuote.premium, moveHomeQuote.deductible!!)
```

Comment:
I wonder if it's possible to have no deductible here, and if we should filter it out earlier, making it not-nullable by putting zero or smth

Reply:
I am not sure when this happens either tbh, and I would be very afraid of assuming a `0` there, as it may somehow be wrong and we might then be showing wrong information instead of having a lack of information.
I just tried to make this cover as much as possible all scenarios that the deductible being nullable makes possible for now. If we figure out there's something wrong with this we can always come back and fix it later